libcxx/test/std/containers/associative/map/map.cons/copy_assign.pass.cpp

```diff
@@ -12,286 +12,251 @@
+      assert(std::next(copy.begin(), 3) == copy.end());
+
+      // Check that orig is still what is expected
+      assert(orig.size() == 3);
+      assert(*std::next(orig.begin(), 0) == V(1, 1));
+      assert(*std::next(orig.begin(), 1) == V(2, 3));
+      assert(*std::next(orig.begin(), 2) == V(3, 6));
+      assert(std::next(orig.begin(), 3) == orig.end());
+    }
+    check_alloc_invariant();
+    { // assign from an empty container into an empty one
+      std::map<int, int, std::less<int>, Alloc> orig(rhs_alloc);
+      std::map<int, int, std::less<int>, Alloc> copy(lhs_alloc);
+      copy = orig;
+      assert(copy.size() == 0);
+      assert(copy.begin() == copy.end());
+
+      // Check that orig is still what is expected
+      assert(orig.size() == 0);
+      assert(orig.begin() == orig.end());
+    }
+    check_alloc_invariant();
+    { // assign from an empty container into a non-empty one
+      using V = std::pair<const int, int>;
+      V arr[] = {V(1, 1), V(2, 3), V(3, 6)};
+      std::map<int, int, std::less<int>, Alloc> orig(rhs_alloc);
+      std::map<int, int, std::less<int>, Alloc> copy(begin(arr), end(arr), rhs_alloc);
+      copy = orig;
+      assert(copy.size() == 0);
+      assert(copy.begin() == copy.end());
+
+      // Check that orig is still what is expected
+      assert(orig.size() == 0);
+      assert(orig.begin() == orig.end());
+    }
   }
 
-  temp1 = ca_allocs;
-  std::sort(temp1.begin(), temp1.end());
-  temp2.clear();
-  std::unique_copy(temp1.begin(), temp1.end(), std::back_inserter<std::vector<int>>(temp2));
-  std::printf("There were %zu different (de)allocators\n", temp2.size());
-
-  for (std::vector<int>::const_iterator it = ca_deallocs.begin(); it != ca_deallocs.end(); ++it) {
-    std::ptrdiff_t const allocs   = std::count(ca_allocs.begin(), ca_allocs.end(), *it);
-    std::ptrdiff_t const deallocs = std::count(ca_deallocs.begin(), ca_deallocs.end(), *it);
-    std::printf("%d: %td vs %td\n", *it, allocs, deallocs);
-    if (allocs != deallocs)
-      return false;
+  {   // Ensure that self-assignment works correctly
+    { // with a non-empty map
+      using V = std::pair<const int, int>;
+      V arr[] = {V(1, 1), V(2, 3), V(3, 6)};
+      std::map<int, int, std::less<int>, Alloc> orig(begin(arr), end(arr), rhs_alloc);
+      orig = orig;
+
+      assert(orig.size() == 3);
+      assert(*std::next(orig.begin(), 0) == V(1, 1));
+      assert(*std::next(orig.begin(), 1) == V(2, 3));
+      assert(*std::next(orig.begin(), 2) == V(3, 6));
+      assert(std::next(orig.begin(), 3) == orig.end());
+    }
+    { // with an empty map
+      std::map<int, int, std::less<int>, Alloc> orig(rhs_alloc);
+      orig = orig;
+
+      assert(orig.size() == 0);
+      assert(orig.begin() == orig.end());
+    }
   }
 
-  return true;
-}
-#endif
+  { // check assignment into a non-empty map
+    check_alloc_invariant();
+    { // LHS already contains, but fewer than needed
```

Comment:
```suggestion
    { // LHS already contains, but fewer than the incoming map
```

Below too.

Otherwise, you're referring too much to your optimization and it doesn't make sense out of context.